experiments/multicluster/controllers/multiclusterlease_controller.go

```diff
@@ -0,0 +1,257 @@
+		}
+		if err := r.stopLeaderElection(mcl); err != nil {
+			return reconcile.Result{}, err
+		}
+		if err := r.removeFinalizer(ctx, mcl); err != nil {
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	}
+
+	if err := r.ensureFinalizer(ctx, mcl); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	v := r.getOrCreateLeaderElector(ctx, mcl)
+	le := v.le
+	lh := v.lh
+	if le.IsLeader() { // leader path
+		if err := le.Renew(ctx); err != nil {
+			return r.handleRenewFailed(ctx, mcl, le, lh)
+		}
+		return r.handleRenewSucceeded(ctx, mcl, le, lh)
+	}
+	if err := le.Acquire(ctx); err != nil { // non-leader path
+		return r.handleAcquireFailed(ctx, mcl, le, lh)
+	}
+	return r.handleAcquireSucceeded(ctx, mcl, le, lh)
+}
+
+func (r *MultiClusterLeaseReconciler) getOrCreateLeaderElector(ctx context.Context, mcl *multiclusterv1alpha1.MultiClusterLease) runningLeaderElector {
+	log := r.log.WithValues("multiClusterUID", mcl.Spec.MultiClusterUID, "identity", mcl.Spec.Identity)
+	r.leaderElectorsMutex.Lock()
+	defer r.leaderElectorsMutex.Unlock()
+
+	if le, found := r.leaderElectors[mapKey(mcl)]; found {
+		return le
+	}
+
+	bl := bucketlease.New(ctx, bucketlease.Config{
+		Identity:      mcl.Spec.Identity,
+		BucketName:    mcl.Spec.Bucket,
+		LeaseName:     mcl.Spec.MultiClusterUID,
+		Log:           log,
+		EventRecorder: r.eventRecorder,
+	})
+
+	le := leaderelection.New(leaderelection.Config{
+		ElectionID:    mcl.Spec.MultiClusterUID,
+		Identity:      mcl.Spec.Identity,
+		Lock:          bl,
+		LeaseDuration: time.Duration(*mcl.Spec.LeaseDurationSeconds) * time.Second,
+		RenewDeadline: time.Duration(*mcl.Spec.RenewDeadlineSeconds) * time.Second,
+		RetryPeriod:   time.Duration(*mcl.Spec.RetryPeriodSeconds) * time.Second,
+		Log:           log,
+		EventRecorder: r.eventRecorder,
+	})
+
+	lh := pauser.New(pauser.PauserConfig{
+		Client:    r.client,
+		Log:       r.log,
+		Namespace: mcl.Namespace,
+		Identity:  mcl.Spec.Identity,
+	})
+	runningLeaderElector := runningLeaderElector{
+		le: le,
+		lh: lh,
+	}
+	r.leaderElectors[mapKey(mcl)] = runningLeaderElector
+	return runningLeaderElector
+}
+
+func (r *MultiClusterLeaseReconciler) handleRenewSucceeded(ctx context.Context, mcl *multiclusterv1alpha1.MultiClusterLease, le *leaderelection.LeaderElector, _ lifecyclehandler.LifecycleHandler) (ctrl.Result, error) {
+	if err := r.updateStatus(ctx, mcl, le.LeaderIdentity(), le.IsLeader()); err != nil {
+		r.log.Error(err, "error updating MultiClusterLease status")
+	}
+	return ctrl.Result{RequeueAfter: time.Duration(*mcl.Spec.RetryPeriodSeconds) * time.Second}, nil
+	// TODO: add jitter
+}
+
+func (r *MultiClusterLeaseReconciler) handleRenewFailed(ctx context.Context, mcl *multiclusterv1alpha1.MultiClusterLease, le *leaderelection.LeaderElector, lh lifecyclehandler.LifecycleHandler) (ctrl.Result, error) {
+	if err := lh.OnStoppedLeading(ctx); err != nil {
+		r.log.Error(err, "error calling stopped leading function")
```

Comment:
I think we have to handle the error, I don't think we can just log it and move on.  (I don't think there's much we can do other than try again, but ... I assume this means that we have two leaders)

Reply:
Agreed. Updated the code so that we retry immediately. Also added a rate limiter to the controller so that we backoff if necessary.